### ISSUE TITLE
TypeScript error using map

### ISSUE THREAD

**Issue by sparkofreason** [fluture-js/Fluture#512]
This issue looks similar to some previous ones involving `fold`. This is using 14.0.0.

 Code:
```
import * as Future from 'fluture'

Future.map(
  Future.chain(
    Future.encase(e => {
      console.log(e)
    })
  )
)
```

Error:

No overload matches this call.
  Overload 1 of 2, '(mapper: never): (source: Functor<unknown>) => Functor<unknown>', gave the following error.
    Argument of type '<LA>(source: FutureInstance<LA, unknown>) => FutureInstance<unknown, void>' is not assignable to parameter of type 'never'.
  Overload 2 of 2, '(mapper: (a: unknown) => FutureInstance<unknown, void>): <F extends Functor<unknown>>(source: F) => ReturnType<(F & { [$T]: FutureInstance<unknown, void>; })["fantasy-land/map"]>', gave the following error.
    Argument of type '<LA>(source: FutureInstance<LA, unknown>) => FutureInstance<unknown, void>' is not assignable to parameter of type '(a: unknown) => FutureInstance<unknown, void>'.
      Types of parameters 'source' and 'a' are incompatible.
        Type 'unknown' is not assignable to type 'FutureInstance<unknown, unknown>'.ts(2769)

**Comment by Avaq** [comment]
Does it still error when you pipe a future into that function?

```js
import * as Future from 'fluture'

Future.resolve(42).pipe(Future.map(
  Future.chain(
    Future.encase(e => {
      console.log(e)
    })
  )
))
```

**Comment by sparkofreason** [comment]
No, interesting. BTW, I took that code from here: https://scrimba.com/scrim/cWWGz7fb

**Comment by Avaq** [comment]
The reason that the original code produces a type error is because TypeScript doesn't have enough information to infer the type of the mapper, which is why the "Overload 1 of 2" shows `never` as the type of `mapper`. By piping a Future into it, the first overload can be properly inferred. If you want to extract that function without using it right away, you'll need to type the arguments yourself, and not use point-free style. Something like:

```ts
const mapChainThing = (arg: Future.FutureInstance<Error, Future.FutureInstance<Error, string>>) => Future.map(...)(arg);
```

FYI the I created the scrimba you linked as part of a discussion I replied to here: https://github.com/fluture-js/Fluture/discussions/478#discussioncomment-978773, in case you want more context.
